crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
+			v.sync()
 		}
 	}
-
-	return false
 }
 
-func (v *validator) updateCRL(endpoint string) error {
-	v.listsLock.RLock()
-	crl, ok := v.lists[endpoint]
-	v.listsLock.RUnlock()
-
-	if !ok || crlHasExpired(crl, nowFunc()) {
-		return v.downloadCRL(endpoint)
+func (v *validator) validatorLoop(ctx context.Context) {
+	var f func()
+	var ok bool
+	for {
+		select {
+		case <-ctx.Done():
+			v.started = false
+			return
+		case f, ok = <-v.crlChan:
+			if ok {
+				f()
+			}
+		}
 	}
-
-	return nil
 }
 
-func (v *validator) verifyCRL(crl *x509.RevocationList) error {
-	v.certificatesLock.RLock()
-	defer v.certificatesLock.RUnlock()
-
-	issuerName := crl.Issuer.String()
-
-	certificate, ok := v.certificates[issuerName]
-	if !ok {
-		return errors.New("CRL signature could not be validated against known certificates")
+func (v *validator) Validate(certificate *x509.Certificate) error {
+	if !v.started {
+		return errors.New("CRL validator is not started")
 	}
-
-	return crl.CheckSignatureFrom(certificate)
+	return v.validateCert(certificate)
 }
 
-func (v *validator) downloadCRL(endpoint string) error {
-	response, err := v.httpClient.Get(endpoint)
-	if err != nil {
-		return err
+func (v *validator) SetValidatePeerCertificateFunc(config *tls.Config) error {
+	if v.started {
+		return errors.New("SetVerifyPeerCertificateFunc must be called before Start")
 	}
 
-	defer response.Body.Close()
-
-	data, err := io.ReadAll(response.Body)
-	if err != nil {
-		return fmt.Errorf("unable to download CRL (url=%s): %w", endpoint, err)
-	}
-
-	crl, err := x509.ParseRevocationList(data)
-	if err != nil {
-		return fmt.Errorf("unable to parse downloaded CRL (url=%s): %w", endpoint, err)
+	parseCertificatesSlice := func(rawCerts [][]byte) ([]*x509.Certificate, error) {
+		var raw []byte
+		for _, rawCert := range rawCerts {
+			raw = append(raw, rawCert...)
+		}
+		return x509.ParseCertificates(raw)
 	}
 
-	if err := v.verifyCRL(crl); err != nil {
-		return fmt.Errorf("CRL verification failed (issuer=%s): %w", crl.Issuer.String(), err)
+	// check that all cert issuers are in the truststore, and add missing crl distribution points.
+	for _, chain := range config.Certificates {
+		certificates, err := parseCertificatesSlice(chain.Certificate)
+		if err != nil {
+			return err
+		}
+		for _, cert := range certificates {
+			issuer := v.truststore[cert.Issuer.String()]
+			if issuer == nil {
+				// This indicates a mismatch between crlValidator truststore and tls.Config. This is a programming error.
+				return errors.New("tls.Config contains certificate from issuer that is not in the truststore: " + cert.Subject.String())
+			}
+			// add any previously unknown CRL distribution points.
+			// (Leaf cert is likely to contain CRLDistributionPoints not found in the truststore.)
+			if len(cert.CRLDistributionPoints) > 0 {
+				for _, endpoint := range cert.CRLDistributionPoints {
+					if v.crls[endpoint] == nil {
+						v.crls[endpoint] = newRevocationList(issuer)
```

Comment:
multiple issuers can use the same CRL distribution endpoint (e.g. QuoVadis PKIo Private intermediate CAs (e.g. Personen, Services) share the same CRL), so I'm missing an append flow here

Reply:
I don't think there can ever be more than one issuer per distribution point. It does not make sense to me to host more than 1 CRL on an endpoint.

All TSP (KPN, QuoVadis, ...) issue their own CRL on a CRLDistributionPoint they control (that is listed on the certs they issue), and the TSP all point to CRL `http://crl.pkioverheid.nl/DomPrivateServicesLatestCRL-G1.crl` that is issued by `Staat der Nederlanden Private Services CA - G1` because that is the issuer of the TSP's.

I think we can safely assume that subjects always point to the CRL of their issuer, because that is the CRL that will revoke them. (Theoretically the spec provides other options, but those do not apply to the pkiOverheid chain and I have not found a chain that deviates from this so far.)

Reply:
Now returns an error when multiple CAs are found for an endpoint

Reply:
I think you're right, I probably misread it (on https://www.quovadisglobal.com/download-roots-crl/#private)